src/cla/tcp.rs

```diff
@@ -0,0 +1,511 @@
+                tokio::spawn(sender_task.run());
+                tokio::spawn(session_task.run());
+                debug!("Spawned TCP session");
+                sessions.insert(self.addr, session_sender);
+                debug!("Started TCP session");
+            }
+            Err(err) => error!("Failed to negotiate session: {}", err),
+        }
+    }
+}
+
+pub struct Listener {
+    tcp_listener: TcpListener,
+    sessions: Arc<Mutex<HashMap<SocketAddr, Sender<(SocketAddr, ByteBuffer)>>>>,
+}
+
+impl Listener {
+    async fn run(self) {
+        loop {
+            match self.tcp_listener.accept().await {
+                Ok((stream, addr)) => {
+                    info!("Incoming connection from: {:?}", addr);
+                    let connection = Connection { stream, addr };
+                    connection.connect(&mut self.sessions.lock().await).await;
+                }
+                Err(e) => {
+                    error!("Couldn't get client: {:?}", e)
+                }
+            }
+        }
+    }
+}
+
+impl TcpConvergenceLayer {
+    pub fn new(port: Option<u16>) -> TcpConvergenceLayer {
+        TcpConvergenceLayer {
+            local_port: port.unwrap_or(4556),
+            sessions: Arc::new(Mutex::new(HashMap::new())),
+            listener: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    async fn send_bundles(
+        &self,
+        addr: &SocketAddr,
+        ready: &[ByteBuffer],
+    ) -> Result<bool, SendError<(SocketAddr, ByteBuffer)>> {
+        if let Some(sender) = self.sessions.lock().await.get(&addr) {
+            debug!("Using existing connection for {}", addr);
+            for bundle in ready {
+                debug!("Sending bundle {:?}", bundle);
+                sender.send((addr.clone(), bundle.clone())).await?
+            }
+            Ok(true)
+        } else {
+            warn!("No connection found for {}", addr);
+            Ok(false)
+        }
+    }
+}
+
+#[derive(Clone, Default, Debug)]
+pub struct TcpConvergenceLayer {
+    local_port: u16,
+    sessions: Arc<Mutex<HashMap<SocketAddr, Sender<(SocketAddr, ByteBuffer)>>>>,
+    listener: Arc<Mutex<Option<JoinHandle<()>>>>,
+}
+
+#[async_trait]
+impl ConvergenceLayerAgent for TcpConvergenceLayer {
+    async fn setup(&mut self) {
+        let tcp_listener = TcpListener::bind(("0.0.0.0", self.local_port))
+            .await
+            .expect("Couldn't create TCP listener");
+        let listener = Listener {
+            tcp_listener,
+            sessions: self.sessions.clone(),
+        };
+        *self.listener.lock().await = Some(tokio::spawn(listener.run()));
+    }
+
+    fn port(&self) -> u16 {
+        self.local_port
+    }
+
+    fn name(&self) -> &'static str {
+        "tcp"
+    }
+
+    async fn scheduled_submission(&self, dest: &str, ready: &[ByteBuffer]) -> bool {
+        let addr = dest.parse().unwrap();
+        let mut session_lock = self.sessions.lock().await;
```

Comment:
can the long holding of this lock be avoided?

Reply:
no, don't think so. adding a new connection should be atomic and during testing releasing the lock in between lead to deadlocks.